src/impl/tpls/KokkosBlas3_gemm_tpl_spec_decl.hpp

```diff
@@ -409,20 +410,29 @@ struct DotBasedGEMM{
     if(numDivPerDot > 1)
       chunkSize++;
 
-    // Initialize C matrix as beta*C
-    // This is not required for the Belos MvTransMv use case which
-    // already initializes C to zero, however, absolutely needed for 
-    // other cases, e.g., when beta is nonzero or when C was not 
-    // initialized for the case beta being zero.  
-    Kokkos::RangePolicy<TagInit, ExecSpace> policy1(0, ndots);
-    Kokkos::parallel_for("Initialize C for Dot Product Based GEMM", policy1, *this);
-
+    // Initialize C matrix if beta != 1
+    if(beta == CVT::zero()) {
+      Kokkos::RangePolicy<TagZero, ExecSpace> policyInit(0, ndots);
+      Kokkos::parallel_for("Initialize C for Dot Product Based GEMM", policyInit, *this);
+    }
+    else if(beta != CVT::one()) {
+      Kokkos::RangePolicy<TagInit, ExecSpace> policyInit(0, ndots);
+      Kokkos::parallel_for("Initialize C for Dot Product Based GEMM", policyInit, *this);
+    }
+    
     // Multiply alpha*A^TB and add it to beta*C
     Kokkos::TeamPolicy<TagMult, ExecSpace> policy2(numTeams, Kokkos::AUTO);
     Kokkos::parallel_for("Perform Dot Product Based GEMM", policy2, *this);
 
   }
 
+  KOKKOS_INLINE_FUNCTION
+  void operator() (const TagZero&, const size_C &i) const {
+    const size_C rowId = i / numCcols;
```

Comment:
Kokkos has MDRangePolicy for multidimensional index ranges.  It looks like that's what you want to do here.

Reply:
If you try MDRangePolicy here make sure to performance test, the matrix shapes (tall skinny) may result in poor performance without tuning tile sizes.

Reply:
@mhoemmen @ndellingwood thanks for useful comments! I will try it.

Reply:
MDRangePolicy results in reduction in **total** runtime of GEMM by **5.95%**, **4.82%**, **0.82%**, and **-1.61%** for 4x4, 8x8, 16x16, and 32x32 C matrices, respectively (averaged over seven different values for the number of rows in A and B in range [10 to 10M]).